---
title: "PCA_Structure_visualization"
author: "Elise Stacy"
date: "2025-01-04"
output: html_document
---
# Load packages
```{r warning = FALSE, message = FALSE}
library(vcfR)
library(adegenet)
library(vegan)
library(pophelper)

library(dplyr)
library(tidyr)
library(reshape2)

library(ggplot2)
library(gridExtra)
library(cowplot)
library(stringr)
```
# Load data
```{r warning = FALSE, message = FALSE}
# All sample data
msat_df <- read.csv("Microsatellites/data/gulo_msat_related_removed.csv")
sample_data <- read.csv("all_samples.csv")

#reformat for genind import
out <- msat_df[1]
out[paste("locus",1:12)] <- lapply(seq(2, ncol(msat_df), 2), function(i) do.call(paste, msat_df[i:(i+1)]))
msat_df_re <- out
 
msat_genind <- df2genind(msat_df_re[2:13], ind.names = msat_df_re$ID, NA.char = "-1", sep =" ")

msat_metadata <- sample_data %>% filter(ID %in% indNames(msat_genind)) %>% arrange(match(ID, indNames(msat_genind)))

msat_genind$pop <- as.factor(msat_metadata$Region)
msat_genpop <- seppop(msat_genind) #separates into populations and orders alphabetically
msat_genind <- repool(msat_genpop) #recombines with new order 

#df <- msat_metadata[match(indNames(msat_genind),msat_metadata$ID),] # add coordinate data - exact location data sensitive, available with error added upon request
#msat_genind$other$lat <- df$Latitude 
#msat_genind$other$long <- df$Longitude 

# SNP data
snp_vcf <- read.vcfR("SNPs/data/AK_YK/051923_iterative_filtering/outliers_removed/gulo_snp_related_removed.vcf")
snp_genind <- vcfR2genind(snp_vcf)

indNames(snp_genind) <- str_extract(indNames(snp_genind), ".*(?=_)")

snp_metadata <- sample_data %>% filter(ID %in% indNames(snp_genind)) %>% arrange(match(ID, indNames(snp_genind)))

snp_genind$pop <- as.factor(snp_metadata$Region)

snp_genpop <- seppop(snp_genind) #separates into populations and orders alphabetically
snp_genind <- repool(snp_genpop) #recombines with new order 
#df <- snp_metadata[match(indNames(snp_genind),snp_metadata$ID),] # add coordinate data - exact location data sensitive, available with error added upon request
#snp_genind$other$lat <- df$Latitude
#snp_genind$other$long <- df$Longitude

# Subset msat data to match snp genotyped samples
msat_sub_genind <- msat_genind[indNames(snp_genind), ]

# For structure, Fst and He ordering
region_order <- c("North Alaska","West Alaska", "Kenai Peninsula","Central Alaska","Northeast Alaska","North Yukon", "Yukon & SE Alaska", "Southeast Alaska")
```
# Population Structure 
## PCA
```{r warning = FALSE, message = FALSE, fig.show="hold", out.width="33%"}
# Create a custom color palette
custom_palette <- c("#393b79","#B89912","#637939","#E5A1CC","#ECDD6C","#AB0478","#8DA8F6","#93E371")

#Microsatellite
msat_df <- tab(msat_genind, freq=TRUE, NA.method="mean") 
msat_pca <- dudi.pca(msat_df, center=TRUE, scale=FALSE,scannf = FALSE, nf = 220)


png(filename="plots/pca_msat.png",
    width = 1200, height = 840, units = "px", pointsize = 32)

# Now use the custom color palette in s.class
s.class(msat_pca$li, fac = pop(msat_genind), xax = 1, yax = 2, col = transp(custom_palette, .7), axesel = FALSE, cstar = 0, cpoint = 2, clabel = 0) 

# Add scatter points for eigenvalues as before
add.scatter.eig(msat_pca$eig[1:20], 3, 1, 2, ratio = .25, "bottomright")
dev.off()

#Microsatellite subset
msat_sub_df <- tab(msat_sub_genind, freq=TRUE, NA.method="mean") 
msat_sub_pca <- dudi.pca(msat_sub_df, center=TRUE, scale=FALSE,scannf = FALSE, nf = 220)


png(filename="plots/pca_msat_sub.png",
    width = 1200, height = 840, units = "px", pointsize = 32)

# Now use the custom color palette in s.class
s.class(msat_sub_pca$li, fac = pop(msat_sub_genind), xax = 1, yax = 2, col = transp(custom_palette, .7), axesel = FALSE, cstar = 0, cpoint = 2, clabel = 0) 

# Add scatter points for eigenvalues as before
add.scatter.eig(msat_sub_pca$eig[1:20], 3, 1, 2, ratio = .25, "bottomright")
dev.off()
# SNP
snp_df <- tab(snp_genind, freq=TRUE, NA.method="mean")
snp_pca <- dudi.pca(snp_df, center=TRUE, scale=FALSE,scannf = FALSE, nf = 220)

png(filename="plots/pca_snp.png",
    width = 1200, height = 840, units = "px", pointsize = 32)

# Now use the custom color palette in s.class
s.class(snp_pca$li, fac = pop(snp_genind), xax = 1, yax = 2, col = transp(custom_palette, .7), axesel = FALSE, cstar = 0, cpoint = 2, clabel = 0) 

# Add scatter points for eigenvalues as before
add.scatter.eig(snp_pca$eig[1:20], 3, 1, 2, ratio = .25, "bottomright")
dev.off()

# Legend
Colors <- c("#AB0478","#93E371","#E5A1CC","#393b79","#B89912","#8DA8F6","#637939","#ECDD6C") #need to be in order of region labels
Regions <- c("Southeast Alaska","South Yukon","North Yukon","Central Alaska","Kenai Peninsula","West Alaska","North Alaska", "Northeast Alaska")
# Create a plot

png(filename="plots/pca_legend.png",
    width = 1200, height = 840, units = "px", pointsize = 32)

plot(1, type = "n", xlab = "", ylab = "")
# Create a custom legend
legend("center", legend = Regions, pch = 19, col = Colors, title = "Regions", pt.cex = 1.5)

dev.off()
```
## Visualize program Structure results
### Read in STRUCTURE results files and create location and sortable individual labels
```{r eval=FALSE}
# Microsats (related removed)
sfiles <- list.files("Structure_results_files/microsats/related_removed/", full.names = T)
slist_msat <- readQ(files=sfiles,filetype="structure")
ind_label <- readQ(files=sfiles[1], indlabfromfile=T) %>% as.data.frame()
ind_label <- tibble::rownames_to_column(ind_label, "ID")

locations_msat <- left_join(ind_label, sample_data)

slist <- slist_msat

# loop thru to apply labels to all structure results
inds_order <- read.csv("Structure_results_files/msat_structure_order.csv")
# add indlab to one run
rownames(slist[[1]]) <- inds_order$x
# if all runs are equal length, add indlab to all runs
if(length(unique(sapply(slist,nrow)))==1) slist <- lapply(slist,"rownames<-",inds_order$x)
# show row names of all runs and all samples
lapply(slist, rownames)[1:200]
slist_msat <- slist

loc_labels_msat <- locations_msat$Region %>% as.data.frame()
colnames(loc_labels_msat) <- "Region"

# Microsats (related kept)
sfiles <- list.files("Structure_results_files/microsats/related_kept/", full.names = T)
slist_msat_r <- readQ(files=sfiles,filetype="structure")
ind_label <- readQ(files=sfiles[1], indlabfromfile=T) %>% as.data.frame()
ind_label <- tibble::rownames_to_column(ind_label, "ID")

locations_msat_r <- left_join(ind_label, sample_data)

slist <- slist_msat_r

# loop thru to apply labels to all structure results
inds_order <- read.csv("Structure_results_files/msat_r_structure_order.csv")
# add indlab to one run
rownames(slist[[1]]) <- inds_order$x
# if all runs are equal length, add indlab to all runs
if(length(unique(sapply(slist,nrow)))==1) slist <- lapply(slist,"rownames<-",inds_order$x)
# show row names of all runs and all samples
lapply(slist, rownames)[1:200]
slist_msat_r <- slist

loc_labels_msat_r <- locations_msat_r$Region %>% as.data.frame()
colnames(loc_labels_msat_r) <- "Region"

# SNPs
sfiles <- list.files("Structure_results_files/snps/", full.names = T)
slist_snp <- readQ(files=sfiles,filetype="structure")
ind_label <- readQ(files=sfiles[1], indlabfromfile=T) %>% as.data.frame()
ind_label <- tibble::rownames_to_column(ind_label, "ID")

locations_snp <- left_join(ind_label, sample_data)

slist <- slist_snp

# loop thru to apply labels to all structure results
inds_order <- read.csv("Structure_results_files/snp_structure_order.csv")
# add indlab to one run
rownames(slist[[1]]) <- inds_order$x
# if all runs are equal length, add indlab to all runs
if(length(unique(sapply(slist,nrow)))==1) slist <- lapply(slist,"rownames<-",inds_order$x)
# show row names of all runs and all samples
lapply(slist, rownames)[1:200]
slist_snp <- slist

loc_labels_snp <- locations_snp$Region %>% as.data.frame()
colnames(loc_labels_snp) <- "Region"
```
### Structure run summaries
```{r warning = FALSE, message = FALSE}
summariseQ(tabulateQ(slist_msat))
summariseQ(tabulateQ(slist_msat_r))
summariseQ(tabulateQ(slist_snp))
```
### Likelihood and Evanno method for choosing optimal K
#### Microsats
##### Related remvoed
```{r warning = FALSE, message = FALSE, fig.show="hold", out.width="50%"}
sr1 <- summariseQ(tabulateQ(slist_msat))
p <- evannoMethodStructure(data=sr1,exportplot=F,returnplot=T,returndata=T,basesize=20,linesize=0.7,  xaxisbreaks = 1:10)

#plot individual plots - likelihood and evanno
plot(p[["plot"]][["grobs"]][[1]])
plot(p[["plot"]][["grobs"]][[4]])
```
##### Related kept
```{r warning = FALSE, message = FALSE, fig.show="hold", out.width="50%"}
sr1 <- summariseQ(tabulateQ(slist_msat_r))
p <- evannoMethodStructure(data=sr1,exportplot=F,returnplot=T,returndata=T,basesize=20,linesize=0.7,  xaxisbreaks = 1:10)

#plot individual plots - likelihood and evanno
plot(p[["plot"]][["grobs"]][[1]])
plot(p[["plot"]][["grobs"]][[4]])
```
#### SNPs
```{r warning = FALSE, message = FALSE, fig.show="hold", out.width="50%"}
sr1 <- summariseQ(tabulateQ(slist_snp))
p <- evannoMethodStructure(data=sr1,exportplot=F,returnplot=T,returndata=T,basesize=20,linesize=0.7,  xaxisbreaks = 1:10)

#plot individual plots - likelihood and evanno
plot(p[["plot"]][["grobs"]][[1]])
plot(p[["plot"]][["grobs"]][[4]])
```
### Microsat and SNP view all reps of K = X
```{r warning = FALSE, message = FALSE}
# Choose dataset to view

#slist <- slist_msat
#loc_labels <- loc_labels_msat

#slist <- slist_msat_r
#loc_labels <- loc_labels_msat_r

slist <- slist_snp
loc_labels <- loc_labels_snp

colors <- c("#C7EDE4","#0A369D", "#EDAE49", "#E0A0D4","#820933", "#120930")

slist1 <- alignK(slist[c(41:60)]) # Choose K value files to view

p1 <- plotQ(slist1, imgoutput="join", returnplot=T, exportplot=F, basesize=5, clustercol = colors, grplab=loc_labels, subsetgrp= region_order, ordergrp = TRUE,grplabsize=2, linesize=0.8, pointsize=4, divcol="white",divtype=1,divsize=0.5, useindlab = T, showindlab = T, returndata = T)

# Within Region, ind sorted by 'label'
p2 <- plotQ(p1$data$qlist,grplab=p1$data$grplab[[1]],imgoutput="join", returnplot=T, exportplot=F, basesize=5, clustercol = colors,ordergrp=F,selgrp="Region", sortind="label",grplabsize=2, linesize=0.8, pointsize=4, divcol="white",divtype=1,divsize=0.5, useindlab = T, showindlab = T, returndata = T)

grid.arrange(p2$plot[[1]])
```

### View supported K values STRUCTURE
#### Microsats
##### Related removed
```{r warning = FALSE, message = FALSE}
slist <- slist_msat
loc_labels <- loc_labels_msat


colors <- c("#FAD188", "#03883f", "#6933A5", "#1FB5D0", "#000002", "#F85B9E")

slist1 <- alignK(slist[c(52,72,113,136)])

#Reordering of cluster labels to keep colors consistent across barplots
slist1$ak_yk.2.2_f # east (cluster 2), west (cluster 1)

# change to east (cluster 1), west (cluster 2)

Cluster1 <- slist1$ak_yk.2.2_f$Cluster2
Cluster2 <- slist1$ak_yk.2.2_f$Cluster1

slist1$ak_yk.2.2_f$Cluster1 <- Cluster1
slist1$ak_yk.2.2_f$Cluster2 <- Cluster2

slist1$ak_yk.3.2_f # S Yukon (cluster 2), Kenai (cluster ?3), west (cluster 1)

#change to S YK (cluster 1), west (cluster 2), Kenai (cluster 3)

Cluster1 <- slist1$ak_yk.3.2_f$Cluster2
Cluster2 <- slist1$ak_yk.3.2_f$Cluster3
Cluster3 <- slist1$ak_yk.3.2_f$Cluster1

slist1$ak_yk.3.2_f$Cluster1 <- Cluster1
slist1$ak_yk.3.2_f$Cluster2 <- Cluster2
slist1$ak_yk.3.2_f$Cluster3 <- Cluster3

slist1$ak_yk.5.20_f # SE (cluster 2) West (cluster 1)

#change to SE (cluster 1), West (cluster 2)

Cluster1 <- slist1$ak_yk.5.20_f$Cluster2
Cluster2 <- slist1$ak_yk.5.20_f$Cluster1
Cluster3 <- slist1$ak_yk.5.20_f$Cluster5
Cluster4 <- slist1$ak_yk.5.20_f$Cluster3
Cluster5 <- slist1$ak_yk.5.20_f$Cluster4


slist1$ak_yk.5.20_f$Cluster1 <- Cluster1
slist1$ak_yk.5.20_f$Cluster2 <- Cluster2
slist1$ak_yk.5.20_f$Cluster3 <- Cluster3
slist1$ak_yk.5.20_f$Cluster4 <- Cluster4
slist1$ak_yk.5.20_f$Cluster5 <- Cluster5

slist1$ak_yk.6.5_f # SE (cluster 1), Kenai (cluster 5)

#change to SE (cluster 1), Kenai (cluster 3)

Cluster1 <- slist1$ak_yk.6.5_f$Cluster1
Cluster2 <- slist1$ak_yk.6.5_f$Cluster2
Cluster3 <- slist1$ak_yk.6.5_f$Cluster5
Cluster4 <- slist1$ak_yk.6.5_f$Cluster3
Cluster5 <- slist1$ak_yk.6.5_f$Cluster4
Cluster6 <- slist1$ak_yk.6.5_f$Cluster6

slist1$ak_yk.6.5_f$Cluster1 <- Cluster1
slist1$ak_yk.6.5_f$Cluster2 <- Cluster2
slist1$ak_yk.6.5_f$Cluster3 <- Cluster3
slist1$ak_yk.6.5_f$Cluster4 <- Cluster4
slist1$ak_yk.6.5_f$Cluster5 <- Cluster5
slist1$ak_yk.6.5_f$Cluster6 <- Cluster6
```

##### Create plot
```{r warning = FALSE, message = FALSE}
p1 <- plotQ(slist1, imgoutput="join", returnplot=T, exportplot=F, basesize=5, clustercol = colors, grplab=loc_labels, subsetgrp= region_order, ordergrp = TRUE,grplabsize=2, linesize=0.8, pointsize=4, divcol="white",divtype=1,divsize=0.5, useindlab = T, showindlab = T, returndata = T)

# Within Region, ind sorted by 'label'
p2 <- plotQ(p1$data$qlist,grplab=p1$data$grplab[[1]],imgoutput="join", returnplot=T, exportplot=F, basesize=10, clustercol = colors,ordergrp=F,selgrp="Region", sortind="label",grplabsize=2, linesize=0.8, pointsize=4, divcol="white",divtype=1,divsize=0.5, useindlab = T, showindlab = F, grplabangle = -8, grplabpos = 0.2, height = 0.75, grplabspacer = -0.1, grplabjust = 0.8, sppos = "left", splabsize = 8, returndata = T)

grid.arrange(p2$plot[[1]])
```
##### Related kept
```{r warning = FALSE, message = FALSE}
slist <- slist_msat_r
loc_labels <- loc_labels_msat_r

colors <- c("#FAD188", "#03883f", "#6933A5", "#1FB5D0", "#000002", "#F85B9E")

#Reordering of cluster labels to keep colors consistent across barplots
slist1 <- alignK(slist[c(52,72,113,136)])

slist1$ak_yk.2.2_f # east (cluster 2), west (cluster 1)

# change to east (cluster 1), west (cluster 2)

Cluster1 <- slist1$ak_yk.2.2_f$Cluster2
Cluster2 <- slist1$ak_yk.2.2_f$Cluster1

slist1$ak_yk.2.2_f$Cluster1 <- Cluster1
slist1$ak_yk.2.2_f$Cluster2 <- Cluster2

slist1$ak_yk.3.2_f # S Yukon (cluster 1), Kenai (cluster 2), west (cluster 3)

#change to SE YK (cluster 1), west (cluster 2), Kenai (cluster 3)

Cluster1 <- slist1$ak_yk.3.2_f$Cluster1
Cluster2 <- slist1$ak_yk.3.2_f$Cluster3
Cluster3 <- slist1$ak_yk.3.2_f$Cluster2

slist1$ak_yk.3.2_f$Cluster1 <- Cluster1
slist1$ak_yk.3.2_f$Cluster2 <- Cluster2
slist1$ak_yk.3.2_f$Cluster3 <- Cluster3

slist1$ak_yk.5.2_f # SE (cluster 4), Kenai (cluster 2)

#change to SE (cluster 1), Kenai (cluster 3)

Cluster1 <- slist1$ak_yk.5.20_f$Cluster4
Cluster2 <- slist1$ak_yk.5.20_f$Cluster1
Cluster3 <- slist1$ak_yk.5.20_f$Cluster2
Cluster4 <- slist1$ak_yk.5.20_f$Cluster3
Cluster5 <- slist1$ak_yk.5.20_f$Cluster5

slist1$ak_yk.5.20_f$Cluster1 <- Cluster1
slist1$ak_yk.5.20_f$Cluster2 <- Cluster2
slist1$ak_yk.5.20_f$Cluster3 <- Cluster3
slist1$ak_yk.5.20_f$Cluster4 <- Cluster4
slist1$ak_yk.5.20_f$Cluster5 <- Cluster5

slist1$ak_yk.6.5_f # SE (cluster 4), Kenai (cluster 2)

#change to SE (cluster 1), Kenai (cluster 3)

Cluster1 <- slist1$ak_yk.6.5_f$Cluster4
Cluster2 <- slist1$ak_yk.6.5_f$Cluster1
Cluster3 <- slist1$ak_yk.6.5_f$Cluster2
Cluster4 <- slist1$ak_yk.6.5_f$Cluster3
Cluster5 <- slist1$ak_yk.6.5_f$Cluster5
Cluster6 <- slist1$ak_yk.6.5_f$Cluster6

slist1$ak_yk.6.5_f$Cluster1 <- Cluster1
slist1$ak_yk.6.5_f$Cluster2 <- Cluster2
slist1$ak_yk.6.5_f$Cluster3 <- Cluster3
slist1$ak_yk.6.5_f$Cluster4 <- Cluster4
slist1$ak_yk.6.5_f$Cluster5 <- Cluster5
slist1$ak_yk.6.5_f$Cluster6 <- Cluster6
```

##### Create plot
```{r warning = FALSE, message = FALSE}
p1 <- plotQ(slist1, imgoutput="join", returnplot=T, exportplot=F, basesize=5, clustercol = colors, grplab=loc_labels, subsetgrp= region_order, ordergrp = TRUE,grplabsize=2, linesize=0.8, pointsize=4, divcol="white",divtype=1,divsize=0.5, useindlab = T, showindlab = T, returndata = T)

# Within Region, ind sorted by 'label'
p2 <- plotQ(p1$data$qlist,grplab=p1$data$grplab[[1]],imgoutput="join", returnplot=T, exportplot=F, basesize=10, clustercol = colors,ordergrp=F,selgrp="Region", sortind="label",grplabsize=2, linesize=0.8, pointsize=4, divcol="white",divtype=1,divsize=0.5, useindlab = T, showindlab = F, grplabangle = -8, grplabpos = 0.2, height = 0.75, grplabspacer = -0.1, grplabjust = 0.8, sppos = "left", splabsize = 8, returndata = T)

grid.arrange(p2$plot[[1]])
```
#### SNPs
```{r warning = FALSE, message = FALSE}
slist <- slist_snp
loc_labels <- loc_labels_snp

colors <- c("#FAD188", "#03883f", "#6933A5", "#1FB5D0", "#000002", "#F85B9E")

slist1 <- alignK(slist[c(52,72,112,132)])

slist1$ak_yk.2.2_f # east (cluster 1), west (cluster 2)

slist1$ak_yk.3.2_f # S Yukon (cluster 3), Kenai (cluster 2), west (cluster 1)

#change to S YK (cluster 1), west (cluster 2), Kenai (cluster 3)

Cluster1 <- slist1$ak_yk.3.2_f$Cluster3
Cluster2 <- slist1$ak_yk.3.2_f$Cluster1
Cluster3 <- slist1$ak_yk.3.2_f$Cluster2

slist1$ak_yk.3.2_f$Cluster1 <- Cluster1
slist1$ak_yk.3.2_f$Cluster2 <- Cluster2
slist1$ak_yk.3.2_f$Cluster3 <- Cluster3


slist1$ak_yk.5.2_f # SE (cluster 2), Yukon (cluster 5), Kenai (cluster 4), West (cluster 1), North (cluster 3)

#change to SE (cluster 1), north (cluster 2), Kenai (cluster 3), Yukon (cluster 4), West (cluster 5)

Cluster1 <- slist1$ak_yk.5.2_f$Cluster3
Cluster2 <- slist1$ak_yk.5.2_f$Cluster1
Cluster3 <- slist1$ak_yk.5.2_f$Cluster2
Cluster4 <- slist1$ak_yk.5.2_f$Cluster5
Cluster5 <- slist1$ak_yk.5.2_f$Cluster4

slist1$ak_yk.5.2_f$Cluster1 <- Cluster1
slist1$ak_yk.5.2_f$Cluster2 <- Cluster2
slist1$ak_yk.5.2_f$Cluster3 <- Cluster3
slist1$ak_yk.5.2_f$Cluster4 <- Cluster4
slist1$ak_yk.5.2_f$Cluster5 <- Cluster5

slist1$ak_yk.6.2_f # SE (cluster 3), Yukon (cluster 5), Kenai (cluster 1), West (cluster 6), North (cluster 4), NE AK (cluster 2)

#change to SE (cluster 1), north (cluster 2), Kenai (cluster 3), Yukon (cluster 4), West (cluster 5), NE AK (6)

Cluster1 <- slist1$ak_yk.6.2_f$Cluster3
Cluster2 <- slist1$ak_yk.6.2_f$Cluster2
Cluster3 <- slist1$ak_yk.6.2_f$Cluster6
Cluster4 <- slist1$ak_yk.6.2_f$Cluster5
Cluster5 <- slist1$ak_yk.6.2_f$Cluster4
Cluster6 <- slist1$ak_yk.6.2_f$Cluster1

slist1$ak_yk.6.2_f$Cluster1 <- Cluster1
slist1$ak_yk.6.2_f$Cluster2 <- Cluster2
slist1$ak_yk.6.2_f$Cluster3 <- Cluster3
slist1$ak_yk.6.2_f$Cluster4 <- Cluster4
slist1$ak_yk.6.2_f$Cluster5 <- Cluster5
slist1$ak_yk.6.2_f$Cluster6 <- Cluster6
```
##### Create plot
```{r}
p1 <- plotQ(slist1, imgoutput="join", returnplot=T, exportplot=F, basesize=5, clustercol = colors, grplab=loc_labels, subsetgrp= region_order, ordergrp = TRUE,grplabsize=2, linesize=0.8, pointsize=4, divcol="white",divtype=1,divsize=0.5, useindlab = T, showindlab = T, returndata = T)

# Within Region, ind sorted by 'label'
p2 <- plotQ(p1$data$qlist,grplab=p1$data$grplab[[1]],imgoutput="join", returnplot=T, exportplot=F, basesize=10, clustercol = colors,ordergrp=F,selgrp="Region", sortind="label",grplabsize=2, linesize=0.8, pointsize=4, divcol="white",divtype=1,divsize=0.5, useindlab = T, showindlab = F, grplabangle = -8, grplabpos = 0.2, height = 0.75, grplabspacer = -0.1, grplabjust = 0.8, sppos = "left", splabsize = 8, returndata = T)

grid.arrange(p2$plot[[1]])
```
